golem-rust/src/transaction.rs

```diff
@@ -0,0 +1,577 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::{Debug, Display, Formatter};
+use std::rc::Rc;
+
+use crate::bindings::golem::api::host::{get_oplog_index, set_oplog_index, OplogIndex};
+use crate::mark_atomic_operation;
+
+/// Represents an atomic operation of the transaction which has a rollback action.
+///
+/// Implement this trait and use it within a `transaction` block.
+/// Operations can also be constructed from closures using `operation`.
+pub trait Operation: Clone {
+    type In: Clone;
+    type Out;
+    type Err;
+
+    /// Executes the operation which may fail with a domain error
+    fn execute(&self, input: Self::In) -> Result<Self::Out, Self::Err>;
+
+    /// Executes a compensation action for the operation. This version has no access to the result
+    /// of the `execute` function, so it can be called in case of compensating advanced, non-domain level errors.
+    ///
+    /// If the operation is only used in `FallibleTransaction`s, this method can be no-op and the actual, result
+    /// dependent compensation can be implemented in `compensate_with_result`.
+    fn compensate(&self, input: Self::In) -> Result<(), Self::Err>;
+
+    /// Executes a compensation action for the operation which ended up with the given result.
+    fn compensate_with_result(
```

Comment:
We can get rid of this variant (and other things with `_with_result` postfix if we make a single `compensate` that gets `result: Option<Result<Out, Err>>` - and it would only be called with `None` in the (not-yet-implemented) advanced case of trying to run compensations after a non-domain level error.